### PR TITLE
Use user's local time settings when displaying times

### LINKED ISSUE #8489: Show times in local user time (such as 12 hour or current timezone) instead of 24 hour
Yesterday PT's forced use of military time caused me to misjudge the time of my exam, and while I was able to reschedule it, this was not a great experience to go through.

In conjunction with [this issue for PrairieTest](https://github.com/PrairieLearn/PrairieTest-feedback/issues/29), we should use the user's system time to show dates instead of defaulting to military time.

Optionally, a setting can be added to allow the user to turn on the current behavior, but as found in a [750 student survey by Prof. Wade, it is clear that users would prefer to have times in the format they prefer rather than an arbitrary format](https://github.com/PrairieLearn/PrairieTest-feedback/issues/29#issuecomment-1572287460).

From a quick glance, it looks like a `credit_date_string` (formatted) is used for student assessments. For access_rules, `access_rule.start_date` and `access_rule.end_date` (formatted differently) is provided. PrairieTest is private but I'd imagine something similar exists, but this issue is only worrying about the PrairieLearn side (but hopes to ease the process of adding this fix there as well).

I'm not sure how easy to change this would be since these dates aren't stored in something like unix time (to my basic knowledge of PL) that could be easily formatted clientside.

I would love to hear any thoughts / alternatives on fixing this issue, as while it is a tiny detail, it definitely has a much larger effect. I'd be happy to write up a PR, but I wanted to get some insight first on the best way to accomplish this fix.

### PR DESCRIPTION
**Pull request by Timothy-Gonzalez**
This is a proof-of-concept PR aimed at displaying times in the user's local time instead of the institution's time format.

Hopefully this work can be continued & ported over to PrairieTest to solve https://github.com/PrairieLearn/PrairieTest-feedback/issues/29!

This PR only affects the admin course instances page.

Before (CST -5:00):
image

After (CST -5:00):
image

After (IRST +3:30):
image

Any feedback is appreciated!

### CODE REVIEW

**Comment by Timothy-Gonzalez** [comment]
I have reviewed and hereby sign the CLA

**Comment by Timothy-Gonzalez** [comment]
recheck

**Comment by jonatanschroeder** [comment]
> Resolves https://github.com/PrairieLearn/PrairieLearn/issues/8489 for PrairieLearn

I don't think it should resolve the issue. I believe it should keep the issue open until this can be done in other pages as well. As such, it addresses the issue, but does not resolve it.

With that in mind: this particular sproc and functionality is used in a lot of places, so there is a lot of potential for this change to break some other pages elsewhere. I believe, in the spirit of what @nwalters512 mentioned in the issue above, it makes more sense to have an initial change be done in a page that has more localized date handling rather than a library used in multiple pages.

I believe there aren't many student pages that fit that criteria, so a staff page may be an option as the initial "guinea pig". Some suggestions:

* the instructor's "Students" tab for an assessment (Date Started column);
* the assessment instance log;
* the issues page (which already has a rudimentary relative date handling);
* the access page for either an assessment or a course instance;
* the course instances page for a course.

That said, this is my personal opinion, I'm happy to defer to Nathan's opinion on the matter. If that is the case:

> * Currently, `credit_date_string` is a string like `100% until DATE`. `credit` is the percent, like `100` in this case. For this to work nicely, we'd probably want to just have `credit_date_string` just be the datetime, but I'm not sure if this will break things. Renaming it to `credit_datetime` would make this change clearer I think.

I think it makes sense to split the `credit_date_string` into `credit` and `date`, and render them both separately.

Regarding `make dev`, I think it makes more sense for you to discuss that in a new GitHub discussion or on Slack, to avoid polluting this PR. I'm assuming you're on Windows (based on your WSL comment), I'm happy to help you in that regard.

**Comment by Timothy-Gonzalez** [comment]
In terms of a guinea pig, I think that makes sense.

Would that mean having this PR just do the minor change on that single page, and then another PR later once everything's verified to work? My original understanding was that this PR would address the whole issue, but demo it out first to prove it works, and then implement the full thing.

For now I'll rewrite this to just be a single change on one of those smaller pages. Let me know what you think makes the most sense. Thanks!

**Comment by jonatanschroeder** [comment]
> Would that mean having this PR just do the minor change on that single page, and then another PR later once everything's verified to work? My original understanding was that this PR would address the whole issue, but demo it out first to prove it works, and then implement the full thing.

Yes, that's what I mean. The sprocs you changed are used in a lot of other locations, which would need to be considered to account for the change in behaviour. This is doable, but I think it makes more sense to have this change be done in a simpler page, gather feedback from users once it's in production (or at least in master), then move on to more complex changes taking the feedback into account. Note that it's always easier to review PRs that make smaller changes instead of a big PR with a tonne of pieces, and whenever possible we've focused on shorter PRs.

**Comment by nwalters512** [comment]
Hey @Timothy-Gonzalez, do you intend to revisit this PR?

**Comment by Timothy-Gonzalez** [comment]
> Hey @Timothy-Gonzalez, do you intend to revisit this PR?

Yes, sorry, been caught with end of year stuff. If this needs to get done sooner feel free to take over.

**Comment by nwalters512** [comment]
@Timothy-Gonzalez checking in again! If you don't plan on working on this, let me know and I'll close out this PR.

I checked on the issue where the time isn't displaying as 24 hour, and it's a known issue with `@github/relative-time-element`: https://github.com/github/relative-time-element/issues/276. Hopefully this will be fixed upstream soon!

**Comment by Timothy-Gonzalez** [comment]
Sorry about how long this took - here's the general summary:
- Added relative-date-time
- Used it in the admin course instances page - works seamlessly
- The content inside is the default fallback if parsing fails - we use a dash in this case (since that's what the current `course-instances.sql` sends)
- Only change required was converted timestamp to use iso8601 so that relative-date-time can process it
- Also - needed to fix the iso8601 proc to include minutes
  - This shouldn't change processing of anything since it's still a valid date - it just explicitly includes minutes now
  - Previously: `2014-04-01T16:30:00-08`
  - Now: `2014-04-01T16:30:00-08:00`
  - With the former, relative-date-time wouldn't accept the date

For future migration to other places, the following steps are probably:
- Change formatted output to iso8601
- Migrate ejs file to use relative-date-time with formatted date as datetime
- None of the js code should need to change

I tested this with various timezones by changing my time and refreshing, and everything seemed to work. I also updated the original PR description to match with these new changes.

Any feedback appreciated, thank you!

File: apps/prairielearn/src/sprocs/format_date_iso8601.sql [hunk 1]
```diff
@@ -5,6 +5,6 @@ CREATE FUNCTION
     ) RETURNS text AS $$
 BEGIN
     EXECUTE 'SET LOCAL timezone TO ' || quote_literal(coalesce(display_timezone, 'UTC'));
-    RETURN to_char(d, 'YYYY-MM-DD') || 'T' || to_char(d, 'HH24:MI:SSOF');
+    RETURN to_char(d, 'YYYY-MM-DD') || 'T' || to_char(d, 'HH24:MI:SSTZH:TZM');
```
Comment (jonatanschroeder) [inline apps/prairielearn/src/sprocs/format_date_iso8601.sql:8]: Be very careful with this change. This sproc is used in a lot of places other than what you are changing.

**Comment by mwest1066** [comment]
To be clear, I see this timezone issue for instructors as completely separate from the 24h vs 12h time display in PrairieTest -- we will absolutely be changing PT to 12h display (in fact a PR is already in the works for that).

**Comment by Timothy-Gonzalez** [comment]
> Be very careful with this change. This sproc is used in a lot of places other than what you are changing.

One solution might be to make another format-date- sproc specifically for relative-time-elements, and then use that to phase out other uses of iso8601.

> Sorry to be so late at commenting on this PR but my first reaction is that we don't want to do this at all. If an instructor is on travel and editing dates for their course, I think they should be viewing and editing them in the timezone of the course, not in the current timezone where the instructor happens to be located.

This probably makes a lot more sense - form instructor's side it should probably always be displayed in institution timezone.

> To be clear, I see this timezone issue for instructors as completely separate from the 24h vs 12h time display in PrairieTest -- we will absolutely be changing PT to 12h display (in fact a PR is already in the works for that).

I don't think instructors should have timezones change for the reasons you explained, unless we explicitly included the timezone - but even then changing behavior after instructors have gotten used to the current one is a easy way to cause issues.

As for PT - 12h is definitely more standard and I'm glad there's a PR in place for that, and this was the original reason for the [previous issue](https://github.com/PrairieLearn/PrairieLearn/issues/8489) and this pr. I still think having nicer, easy to parse dates that `relative-time-element` provides (`Tue, Apr 9, 07:01 PM` vs `2024-03-09 19:01:00 (CDT)`) would also be a welcome change. Comparing numbers to actual days when trying to schedule things can make for easy off-by-one errors.

----

I think changing to a specific user's time display makes much more sense from a student perspective. As a student, I'd expect times to be in a non-iso format that's easy to parse & in my current timezone - not whatever my school sets as a setting.

So - this kind of is a roadblock - we want a change that doesn't affect many people to demo this out (instructors normally being a good candidate for this since those pages are much less traffic) - but it doesn't make sense to change for instructors. I'm not sure what the path forward is.

If you want to handle this internally - that's fine too - just let me know. Thanks!

**Comment by nwalters512** [comment]
Apologies, I should have made sure that #8489 spelled things out more explicitly before suggesting you open a PR. I also should have made sure other core devs were aligned.

What I was trying to recommend in the discussion on that issue was client-side formatting specifically just for 12 vs 24 hour formatting. This would not initially include timezone localization, relative formatting, or anything else one might use client side formatting for.

@mwest1066 want to discuss a little more on the original issue?

**Comment by Timothy-Gonzalez** [comment]
Based on [this reply](https://github.com/PrairieLearn/PrairieLearn/issues/8489#issuecomment-2024201791):
> I also think that PT and PL may benefit from different choices on some of these issues.
> 
> We have a significant change to PT that changes the date/time display in a number of ways. This will be deployed shortly. My current suggestion is to wait for that and see how it plays out, and then return to the question of date display in PL after we've collected more experience in PT.

I'm going to close this PR, and once it's reassessed what is needed for the original issue, we can create a new PR if needed.

**Pull request closed without merging**